core/statedb/statedb.go

```diff
@@ -437,44 +437,92 @@ func (s *StateDB) PrepareNft(nftIndex int64) error {
 	return nil
 }
 
-func (s *StateDB) UpdateAccountTree(accounts []int64, assets []int64) error {
-	for _, accountIndex := range accounts {
-		for _, assetId := range assets {
-			assetLeaf, err := tree.ComputeAccountAssetLeafHash(
-				s.AccountMap[accountIndex].AssetInfo[assetId].Balance.String(),
-				s.AccountMap[accountIndex].AssetInfo[assetId].LpAmount.String(),
-				s.AccountMap[accountIndex].AssetInfo[assetId].OfferCanceledOrFinalized.String(),
-			)
-			if err != nil {
-				return fmt.Errorf("compute new account asset leaf failed: %v", err)
-			}
-			err = s.AccountAssetTrees[accountIndex].Set(uint64(assetId), assetLeaf)
-			if err != nil {
-				return fmt.Errorf("update asset tree failed: %v", err)
+func (s *StateDB) IntermediateRoot(cleanDirty bool) error {
+	for accountIndex, assetsMap := range s.DirtyAccountsAndAssetsMap {
+		assets := make([]int64, 0, len(assetsMap))
+		for assetIndex, isDirty := range assetsMap {
+			if !isDirty {
+				continue
 			}
+			assets = append(assets, assetIndex)
+		}
+
+		err := s.updateAccountTree(accountIndex, assets)
```

Comment:
For the failed tx, it will also mark the dirty map, will it cause problem?

Reply:
It won't, the failed tx only update the executor's dirty map, it won't sync to the state cache's dirty map.